_sips/sips/2019-10-07-preprocessor-for-conditional-compilation.md

```diff
@@ -0,0 +1,416 @@
+
+The new preprocessor syntax is not 100% compatible with existing Scala code but the impact is low. It requires a rather contrived case of a type projection split across multiple lines to construct a conflict:
+
+```scala
+trait A { type B = Int }
+
+val (x: A
+          #B) = 1
+```
+
+In this example `#B` would be parsed as an unrecognized preprocessor directive, leading to a syntax error. Triggering the preprocessor can be avoided with additional whitespace after the `#`:
+
+```scala
+trait A { type B = Int }
+
+val (x: A
+          # B) = 1
+```
+
+Since preprocessor parsing is performed after Scala tokenization, a preprocessor directive can only occur *between* tokens but not within a token. In particular, preprocessor directives are not recognized within comments (including scaladoc comments) and multi-line string literals.
+
+## Implementation
+
+The preprocessor should be shipped in Scala 2.14 so that it may be used for the transition from 2.14 to 3.0. Backporting to 2.13.x and 2.12.x (behind a compiler flag) is also desirable and should be straight-forward.
+
+A prototype implementation based on Scala 2.13.0 which contains both the preprocessor described so far and the alternative implementation below can be found [in this branch](https://github.com/szeiger/scala/tree/wip/preprocessor). Both implementations can be used simultaneously.
+
+The preprocessor is part of the scanner, filtering the tokens that are ermitted from it while keeping the parser's state consistent. At the same time it requires access to the parser for parsing preprocessor expressions. This can be achieved with an abstract expression parsing method in the scanner that is implemented by the parser. This means that it is no longer possible to instantiate a scanner without a parser but the scanner can still be used directly (after instantiating parser plus scanner). This change did not cause any problems and required only a few small changes in the prototype implementation.
+
+## Drawbacks
+
+### Tooling
+
+Third-party tooling based on the standard Scala parser is expected to continue to work without major changes. Custom parser implementations need to be updated to support the preprocessor syntax. Ignoring the preprocessor after parsing is a sensible default in many cases. For example, an IDE could treat any omitted code block like a comment. This is similar to the current situation with separate source directories where only one set of directories, corresponding to the main build version, contains the sources on which the IDE operates.
+
+Some source-based tools will face bigger difficulties. In particular, source rewriting becomes much more challenging in the presence of conditional code blocks, especially if they do not have to conform to the AST's nesting structure.
```

Comment:
IMHO most of these fall under "continue to work without major changes". Third-party parsers, syntax highlighters and IDEs need to support the new syntax but are not expected to work on multiple conditional versions at the same time. I hacked the parser and interpreter for scalac together in a few days (without much previous experience working on the scalac parser) so I don't expect this to be a big problem.

The configuration parameters determine which code blocks are active. Any inactive block can be treated the same way as a comment. As mentioned in the first paragraph, this gives you the same level of support that you would get with the current directory-based approach. Anything beyond that would be nice to have, but is not a requirement for supporting the new conditional compilation scheme.

The situation for semantic refactoring tools is similiar: they could work on one conditional version, all others would be considered comments and run the risk of getting out of sync. You already have the same limitation with directory-based conditional compilation.

Code formatting tools are affected the most because they work well with directory-based conditional compilation. You would have to do multiple passes over a file for the different configurations. Things may get more difficult if reformatting needs to move things around across multiple lines.